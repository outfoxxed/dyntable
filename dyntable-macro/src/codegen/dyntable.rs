use proc_macro2::TokenStream;
use quote::format_ident;
use syn::{Signature, Token};

use crate::parse::{DynTraitInfo, VTableEntry};

pub fn impl_dyntable(dyntrait: &DynTraitInfo) -> TokenStream {
	let proxy_trait = format_ident!("__DynTable_{}", dyntrait.dyntrait.ident);
	let ident = &dyntrait.dyntrait.ident;
	let vtable_ident = &dyntrait.vtable.name;
	let (_, ty_generics, where_clause) = dyntrait.generics.split_for_impl();
	let impl_generics = dyntrait.generics.params.clone().into_iter()
    .collect::<Vec<_>>();
	let where_entries = where_clause
    .into_iter()
    .flat_map(|clause| &clause.predicates);

	let vtable_entries = dyntrait.entries.iter()
    .map(|entry| match entry {
			VTableEntry::Subtable(subtable) => {
				let entry_name = &subtable.ident;
				quote::quote! {
					#entry_name: <__DynTarget as ::dyntable::DynTable<
						<(dyn #ident #ty_generics + 'static) as ::dyntable::VTableRepr>::VTable,
					>>::VTABLE
				}
			},
			VTableEntry::Method(method) => {
				let Signature {
					unsafety,
					abi,
					fn_token,
					ident,
					inputs,
					variadic,
					output,
					..
				} = &method.sig;

				let inputs = inputs.iter()
					.map(|_| <Token![_]>::default());

				let output = match output {
					syn::ReturnType::Default => None,
					syn::ReturnType::Type(..) => Some(<Token![_]>::default()),
				}.into_iter();

				quote::quote! {
					#ident: unsafe {
						::core::intrinsics::transmute(
							__DynTarget::#ident as
								#unsafety #abi #fn_token (
									#(#inputs,)* #variadic
								) #( -> #output)*
						)
					}
				}
			},
		});

	quote::quote! {
		#[allow(non_camel_case_types)]
		unsafe trait #proxy_trait<'v, V: 'v + ::dyntable::VTable> {
			const VTABLE: V;
			const STATIC_VTABLE: &'v V;
		}

		unsafe impl<
			'__dyn_vtable,
			#(#impl_generics + '__dyn_vtable,)*
			__DynTarget,
		> ::dyntable::__private::DynTable2<'__dyn_vtable, #vtable_ident #ty_generics>
		for ::dyntable::__private::DynImplTarget<__DynTarget, #vtable_ident #ty_generics>
		where
			#(#where_entries,)*
			__DynTarget: #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>,
		{
			const VTABLE: #vtable_ident #ty_generics = __DynTarget::VTABLE;
			const STATIC_VTABLE: &'__dyn_vtable #vtable_ident #ty_generics = __DynTarget::STATIC_VTABLE;
		}

		unsafe impl<
			'__dyn_vtable,
			#(#impl_generics + '__dyn_vtable,)*
			__DynTarget: #ident #ty_generics,
		> #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>
		for __DynTarget
		#where_clause {
			const STATIC_VTABLE: &'__dyn_vtable #vtable_ident #ty_generics =
				&<Self as #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>>::VTABLE;
			const VTABLE: #vtable_ident #ty_generics = #vtable_ident {
				#(#vtable_entries,)*
			};
		}
	}
}
