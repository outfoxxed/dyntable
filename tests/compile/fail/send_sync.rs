use dyntable::{DynBox, dyntable};

#[dyntable]
trait TestTrait {}

fn require_send<T: Send>(_: T) {}
fn require_sync<T: Sync>(_: T) {}

struct Dummy;
impl TestTrait for Dummy {}

fn main() {
	let dynbox = DynBox::<dyn TestTrait>::new(Dummy);
	let dynbox2 = DynBox::<dyn TestTrait>::new(Dummy);

	// &Dyn<dyn TestTrait> should not be Send
	require_send(&*dynbox);
	// &Dyn<dyn TestTrait> should not be Sync
	require_sync(&*dynbox);

	// &Dyn<dyn TestTrait> should not be Send
	require_send(dynbox);
	// &Dyn<dyn TestTrait> should not be Sync
	require_sync(dynbox2);
}
