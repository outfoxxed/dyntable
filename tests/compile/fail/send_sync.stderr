error[E0277]: `(dyn NoBounds + 'static)` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:23:15
   |
23 |     require_send(DynBox::borrow_mut(&mut dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `(dyn NoBounds + 'static)` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sync` is not implemented for `(dyn NoBounds + 'static)`
   = note: required for `DynRefMut<'_, dyn TestTrait>` to implement `Send`
note: required by a bound in `require_send`
  --> tests/compile/fail/send_sync.rs:11:20
   |
11 | fn require_send<T: Send>(_: T) {}
   |                    ^^^^ required by this bound in `require_send`

error[E0277]: `*mut c_void` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:25:15
   |
25 |     require_sync(DynBox::borrow_mut(&mut dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*mut c_void` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `DynRefMut<'_, dyn TestTrait>`, the trait `Sync` is not implemented for `*mut c_void`
   = note: required because it appears within the type `DynPtr<dyn TestTrait>`
   = note: required because it appears within the type `DynRefMut<'_, dyn TestTrait>`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`

error[E0277]: `*const TestTraitVTable` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:25:15
   |
25 |     require_sync(DynBox::borrow_mut(&mut dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*const TestTraitVTable` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `DynRefMut<'_, dyn TestTrait>`, the trait `Sync` is not implemented for `*const TestTraitVTable`
   = note: required because it appears within the type `DynPtr<dyn TestTrait>`
   = note: required because it appears within the type `DynRefMut<'_, dyn TestTrait>`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`

error[E0277]: `(dyn NoBounds + 'static)` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:28:15
   |
28 |     require_send(DynBox::borrow(&dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^ `(dyn NoBounds + 'static)` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sync` is not implemented for `(dyn NoBounds + 'static)`
   = note: required for `DynRef<'_, dyn TestTrait>` to implement `Send`
note: required by a bound in `require_send`
  --> tests/compile/fail/send_sync.rs:11:20
   |
11 | fn require_send<T: Send>(_: T) {}
   |                    ^^^^ required by this bound in `require_send`

error[E0277]: `*mut c_void` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:30:15
   |
30 |     require_sync(DynBox::borrow(&dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^ `*mut c_void` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `DynRef<'_, dyn TestTrait>`, the trait `Sync` is not implemented for `*mut c_void`
   = note: required because it appears within the type `DynPtr<dyn TestTrait>`
   = note: required because it appears within the type `DynRef<'_, dyn TestTrait>`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`

error[E0277]: `*const TestTraitVTable` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:30:15
   |
30 |     require_sync(DynBox::borrow(&dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^ `*const TestTraitVTable` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `DynRef<'_, dyn TestTrait>`, the trait `Sync` is not implemented for `*const TestTraitVTable`
   = note: required because it appears within the type `DynPtr<dyn TestTrait>`
   = note: required because it appears within the type `DynRef<'_, dyn TestTrait>`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`

error[E0277]: `(dyn NoBounds + 'static)` cannot be sent between threads safely
  --> tests/compile/fail/send_sync.rs:33:15
   |
33 |     require_send(dynbox);
   |     ------------ ^^^^^^ `(dyn NoBounds + 'static)` cannot be sent between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Send` is not implemented for `(dyn NoBounds + 'static)`
   = note: required for `DynBox<dyn TestTrait>` to implement `Send`
note: required by a bound in `require_send`
  --> tests/compile/fail/send_sync.rs:11:20
   |
11 | fn require_send<T: Send>(_: T) {}
   |                    ^^^^ required by this bound in `require_send`

error[E0277]: `(dyn NoBounds + 'static)` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:35:15
   |
35 |     require_sync(dynbox2);
   |     ------------ ^^^^^^^ `(dyn NoBounds + 'static)` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sync` is not implemented for `(dyn NoBounds + 'static)`
   = note: required for `DynBox<dyn TestTrait>` to implement `Sync`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`

error[E0277]: `(dyn Send + 'static)` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:42:15
   |
42 |     require_send(DynBox::borrow_mut(&mut dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `(dyn Send + 'static)` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sync` is not implemented for `(dyn Send + 'static)`
   = note: required for `DynRefMut<'_, dyn TestSend>` to implement `Send`
note: required by a bound in `require_send`
  --> tests/compile/fail/send_sync.rs:11:20
   |
11 | fn require_send<T: Send>(_: T) {}
   |                    ^^^^ required by this bound in `require_send`

error[E0277]: `(dyn Send + 'static)` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:43:15
   |
43 |     require_send(DynBox::borrow(&dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^ `(dyn Send + 'static)` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sync` is not implemented for `(dyn Send + 'static)`
   = note: required for `DynRef<'_, dyn TestSend>` to implement `Send`
note: required by a bound in `require_send`
  --> tests/compile/fail/send_sync.rs:11:20
   |
11 | fn require_send<T: Send>(_: T) {}
   |                    ^^^^ required by this bound in `require_send`

error[E0277]: `*mut c_void` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:44:15
   |
44 |     require_sync(DynBox::borrow_mut(&mut dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*mut c_void` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `DynRefMut<'_, dyn TestSend>`, the trait `Sync` is not implemented for `*mut c_void`
   = note: required because it appears within the type `DynPtr<dyn TestSend>`
   = note: required because it appears within the type `DynRefMut<'_, dyn TestSend>`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`

error[E0277]: `*const TestSendVTable` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:44:15
   |
44 |     require_sync(DynBox::borrow_mut(&mut dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*const TestSendVTable` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `DynRefMut<'_, dyn TestSend>`, the trait `Sync` is not implemented for `*const TestSendVTable`
   = note: required because it appears within the type `DynPtr<dyn TestSend>`
   = note: required because it appears within the type `DynRefMut<'_, dyn TestSend>`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`

error[E0277]: `*mut c_void` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:45:15
   |
45 |     require_sync(DynBox::borrow(&dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^ `*mut c_void` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `DynRef<'_, dyn TestSend>`, the trait `Sync` is not implemented for `*mut c_void`
   = note: required because it appears within the type `DynPtr<dyn TestSend>`
   = note: required because it appears within the type `DynRef<'_, dyn TestSend>`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`

error[E0277]: `*const TestSendVTable` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:45:15
   |
45 |     require_sync(DynBox::borrow(&dynbox));
   |     ------------ ^^^^^^^^^^^^^^^^^^^^^^^ `*const TestSendVTable` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: within `DynRef<'_, dyn TestSend>`, the trait `Sync` is not implemented for `*const TestSendVTable`
   = note: required because it appears within the type `DynPtr<dyn TestSend>`
   = note: required because it appears within the type `DynRef<'_, dyn TestSend>`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`

error[E0277]: `(dyn Send + 'static)` cannot be shared between threads safely
  --> tests/compile/fail/send_sync.rs:48:15
   |
48 |     require_sync(dynbox);
   |     ------------ ^^^^^^ `(dyn Send + 'static)` cannot be shared between threads safely
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Sync` is not implemented for `(dyn Send + 'static)`
   = note: required for `DynBox<dyn TestSend>` to implement `Sync`
note: required by a bound in `require_sync`
  --> tests/compile/fail/send_sync.rs:12:20
   |
12 | fn require_sync<T: Sync>(_: T) {}
   |                    ^^^^ required by this bound in `require_sync`
